core/src/main/java/org/apache/iceberg/variants/VariantBuilderBase.java

```diff
@@ -0,0 +1,504 @@
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+abstract class VariantBuilderBase {
+  protected static final int MAX_SHORT_STR_SIZE = 0x3F;
+
+  private final ByteBufferWrapper buffer;
+  private final Dictionary dict;
+  private int startPos;
+
+  VariantBuilderBase(ByteBufferWrapper buffer, Dictionary dict) {
+    this.buffer = buffer;
+    this.dict = dict;
+    startPos = buffer.pos;
+  }
+
+  /**
+   * Builds the variant metadata from `dictionaryKeys` and returns the resulting Variant object.
+   *
+   * @return The constructed Variant object.
+   */
+  public Variant build() {
+    int numKeys = dict.size();
+
+    // Calculate total size of dictionary strings
+    long numStringBytes = dict.totalBytes();
+    if (numStringBytes > VariantConstants.SIZE_LIMIT) {
+      throw new VariantSizeLimitException();
+    }
+
+    // Determine the number of bytes required for dictionary size and offset entry
+    int offsetSize = sizeOf(Math.max((int) numStringBytes, numKeys));
+
+    // metadata: header byte, dictionary size, offsets and string bytes
+    long metadataSize = 1 + offsetSize + (numKeys + 1) * offsetSize + numStringBytes;
+
+    // Ensure the metadata size is within limits
+    if (metadataSize > VariantConstants.SIZE_LIMIT) {
+      throw new VariantSizeLimitException();
+    }
+
+    ByteBufferWrapper metadataBuffer =
+        new ByteBufferWrapper((int) metadataSize, (int) metadataSize);
+
+    // Write header byte (version + offset size)
+    metadataBuffer.addByte(VariantUtil.metadataHeader(VariantConstants.VERSION, offsetSize));
+
+    // Write number of keys
+    metadataBuffer.writeLittleEndianUnsigned(numKeys, offsetSize);
+
+    // Write offsets
+    int currentOffset = 0;
+    for (byte[] key : dict.getKeys()) {
+      metadataBuffer.writeLittleEndianUnsigned(currentOffset, offsetSize);
+      currentOffset += key.length;
+    }
+    metadataBuffer.writeLittleEndianUnsigned(numStringBytes, offsetSize);
+
+    // Write dictionary strings
+    dict.getKeys().forEach(metadataBuffer::addBytes);
+
+    return new VariantImpl(metadataBuffer.toByteArray(), buffer.toByteArray());
+  }
+
+  protected void writeNullInternal() {
+    buffer.addByte(VariantUtil.primitiveHeader(Variants.Primitives.TYPE_NULL));
+  }
+
+  protected void writeBooleanInternal(boolean value) {
+    buffer.addByte(
+        VariantUtil.primitiveHeader(
+            value ? Variants.Primitives.TYPE_TRUE : Variants.Primitives.TYPE_FALSE));
+  }
+
+  /**
+   * Writes a numeric value to the variant builder, automatically choosing the smallest type (INT8,
+   * INT16, INT32, or INT64) to store the value efficiently.
+   *
+   * @param value The numeric value to append.
+   */
+  protected void writeNumericInternal(long value) {
```

Comment:
Why does this duplicate all of the serialization that is already implemented in `PrimitiveWrapper`?

Reply:
I added `writePrimitive`() in `ByteBufferWrapper` which will call the serialization in `PrimitiveWrapper`.